docs/source/app_developers_guide/docker.rst

```diff
@@ -2,696 +2,672 @@
 
-  $ docker exec -it sawtooth-settings-tp-default bash
+.. _container-names-label:
 
-To see which components are running, run this command from the container:
+Step 8: Connect to Each Container (Optional)
+============================================
 
-.. code-block:: console
+Use this information when you need to connect to a Docker container in this
+development environment. For example, you can examine the log files or check the
+status of Sawtooth components in any container.
 
-  $ ps --pid 1 fw
-    PID TTY      STAT   TIME COMMAND
-    1 ?        Ssl    0:00 /usr/bin/python3 /usr/bin/settings-tp -vv tcp://validator:4004
+ #. Use the following ``docker exec`` command from your host system to connect to
+    a Sawtooth Docker container.
 
+    .. code-block:: console
 
-The IntegerKey Transaction Processor Container
-----------------------------------------------
+       $ docker exec -it {ContainerName} bash
 
-* Runs a single IntegerKey transaction processor
-* Handles transactions of the IntegerKey transaction family
-* Hostname: ``intkey-tp-python``
-* Container name: ``sawtooth-intkey-tp-python-default``
+    The Docker Compose file defines the name of each container. It also specifies
+    the TCP port and host name, if applicable. The following table shows the values
+    in the example Compose file, ``sawtooth-default.yaml``.
 
-Log into this container by running this command from the host computer's
-terminal window:
+    +---------------+---------------------------------------+----------+----------------------+
+    | **Component** | **Container Name**                    | **Port** | **Host Name**        |
+    +===============+=======================================+==========+======================+
+    | Validator     | ``sawtooth-validator-default``        | 4004     | ``validator``        |
+    +---------------+---------------------------------------+----------+----------------------+
+    | REST API      | ``sawtooth-rest-api-default``         | 8008     |                      |
+    +---------------+---------------------------------------+----------+----------------------+
+    | Settings TP   | ``sawtooth-settings-tp-default``      |          | ``settings-tp``      |
+    +---------------+---------------------------------------+----------+----------------------+
+    | IntegerKey TP | ``sawtooth-intkey-tp-python-default`` |          | ``intkey-tp-python`` |
+    +---------------+---------------------------------------+----------+----------------------+
+    | XO TP         | ``sawtooth-xo-tp-python-default``     |          | ``xo-tp-python``     |
+    +---------------+---------------------------------------+----------+----------------------+
+    | Shell         | ``sawtooth-shell-default``            |          |                      |
+    +---------------+---------------------------------------+----------+----------------------+
```

Comment:
There is some nuances about the context in which these values have meaning. To make sense of it, you need to separate out the following:
- What you can see when interacting with the containers from the host environment
- What you can see when interacting with the other containers from within a container

From the host environment, you can refer to each of the containers with name listed under "Container Name". You cannot, however, interact with the containers using their "Host Name" from the host. So you can do things like `docker exec -it sawtooth-validator-default bash` from the host, but not things like `curl http://rest-api:8008` from the host.

The "Host Name" is only valid when containers to refer to each other (this has to do with virtual docker networks and special DNS resolution rules within containers). So if you were to connect to a container with `docker exec -it sawtooth-validator-default bash` and then _from within the container_ do `curl http://rest-api:8008`, it would work.

Perhaps the trickiest part is the ports. There is a difference between "exposed" ports and "forwarded" ports and the naming conventions are not helpful. An "exposed" port is a port that can be seen from one container to another container, but not from the host to the container. On the other hand, a "forwarded" port is one that can be seen from the host to the container, but not necessarily from one container to another. To make this even more confusing, in the `sawtooth-default.yaml` container being referenced, within the validator and the rest api, the port mentioned is both exposed and forwarded:

```
    expose: # These are the exposed ports (container-to-container)
      - 4004
    ports: # These are the forwarded ports (host-to-container)
      - "4004:4004"
```

Reply:
Thanks for the explanation.  I propose adding this sentence after the table to explain about the exposed and forwarded/published ports: 

     "The validator and REST API ports are exported to other containers and
       forwarded (published) for external connections, such as from your host
       system."

Reply:
Did you mean to say "exported" here? Or should this say "exposed"?

Reply:
Fixed.